ott/core/initializers.py

```diff
@@ -278,6 +285,232 @@ def init_dual_a(
+      self, ot_prob: linear_problems.LinearProblem, lse_mode: bool
+  ) -> jnp.ndarray:
+    # Detect if the problem is batched.
+    if ot_prob.a.ndim == 1:
+      vmap_a_val = None
+    else:
+      assert ot_prob.a.ndim == 2
+      vmap_a_val = 0
+
+    if ot_prob.b.ndim == 1:
+      vmap_b_val = None
+    else:
+      assert ot_prob.b.ndim == 2
+      vmap_b_val = 0
+
+    if vmap_a_val is not None or vmap_b_val is not None:
+      compute_f_maybe_batch = jax.vmap(
+          self._compute_f, in_axes=(vmap_a_val, vmap_b_val, None)
+      )
+    else:
+      compute_f_maybe_batch = self._compute_f
+
+    init_f = compute_f_maybe_batch(ot_prob.a, ot_prob.b, self.state.params)
+    f_u = init_f if lse_mode else ot_prob.geom.scaling_from_potential(init_f)
+    return f_u
+
+  def _get_update_fn(self):
+    """Return the implementation (and jitted) update function."""
+
+    def dual_obj_loss_single(params, a, b):
+      r"""Compute the loss for a single coupling problem.
+
+      The loss :math:`J` is the negated entropy-regularized objective and
+      is given by
+
+      .. math::
+        -J(f; \alpha, \beta, c) := \langle f, a\rangle + \langle g, b \rangle - \epsilon\left\langle \exp\{f/\epsilon\}, K\exp\{g/\epsilon\}\right\rangle
+
+      where :math:`K_{i,j} := -C_{i,j}/\\epsilon` is the *Gibbs kernel*.
+
+      Args:
+        a: Probabilites of the :math:`\alpha` measure's atoms
+        b: Probabilites of the :math:`\beta` measure's atoms
+        params: The parameters of the Meta model.
+
+      Returns:
+        The dual loss (the negated dual objective)
+      """
+      f_pred = self._compute_f(a, b, params)
+      g_pred = self.geom.update_potential(
+          f_pred, jnp.zeros_like(b), jnp.log(b), 0, axis=0
+      )
+      g_pred = jnp.where(jnp.isfinite(g_pred), g_pred, 0.)
+
+      ot_prob = linear_problems.LinearProblem(geom=self.geom, a=a, b=b)
+      dual_obj = sinkhorn.ent_reg_cost(f_pred, g_pred, ot_prob, lse_mode=True)
+      loss = -dual_obj
+      return loss, f_pred
+
+    def loss_batch(params, a, b):
+      loss_fn = functools.partial(dual_obj_loss_single, params=params)
+      loss, f_pred = jax.vmap(loss_fn)(a=a, b=b)
+      return jnp.mean(loss), f_pred
+
+    @jax.jit
+    def update(state, a, b):
+      if a.ndim == 1:
+        a = jnp.expand_dims(a, 0)
+
+      if b.ndim == 1:
+        b = jnp.expand_dims(b, 0)
+
+      grad_fn = jax.value_and_grad(loss_batch, has_aux=True)
+      (loss, init_f), grads = grad_fn(state.params, a, b)
+      return loss, init_f, state.apply_gradients(grads=grads)
+
+    return update
+
+  def _compute_f(self, a, b, params):
+    r"""Predict the optimal :math:`f` potential.
+
+    Args:
+      a: Probabilites of the :math:`\alpha` measure's atoms
+      b: Probabilites of the :math:`\beta` measure's atoms
+      params: The parameters of the Meta model.
+
+    Returns:
+      The :math:`f` potential
+    """
+    return self.meta_model.apply({'params': params}, a, b)
+
```

Comment:
There's missing the implementation of `tree_flatten`, should be something like (am not sure about the `rng`):
```python
def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:
  return [self.geom, self.meta_model, self.opt], {'rng': self.rng}
```

Reply:
I've added this, and also included the training state (which I've added as an optional argument to the class):

```Python
def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:                                       
    return [self.geom, self.meta_model, self.opt], {'rng': self.rng, 'state': self.state}
```

And I quickly tested it with:

```Python
leaves, treedef = jax.tree_util.tree_flatten(meta_initializer)
jax.tree_util.tree_unflatten(treedef, leaves)
```